configdialog.cpp

```diff
@@ -488,6 +529,79 @@ void PaletteCustomizeDialog::delete_current_palette() {
 	}
 }
 
+
+void PaletteCustomizeDialog::do_png_export(const std::wstring &base_path, const CurrentPalette &palette) {
+	Gdiplus::GdiplusStartupInput gdiStartup;
+	ULONG_PTR gdiplus_token;
+	Gdiplus::GdiplusStartup(&gdiplus_token, &gdiStartup, NULL);
+
+	for (const auto &bitmap_def : Bitmaps::All) {
+		// The most reliable way to get the data for the bitmap to GDI+ is to
+		// use the expanded data in its full, unindexed, 32 bits per pixel glory.
+		// A lot like how it is for OpenGL!
+		auto texture = Texture::of(&palette.data, bitmap_def);
+		GLubyte *texture_data = texture->data();
+
+		// We store texture data as RGBA, but GDI+ wants them as BGRA.
+		// So, swap the texture's red and blue channels.
+		for (int i = 0; i < BITMAP_WH * BITMAP_WH; i++) {
+			auto red = texture_data[i * 4 + RED];
+			texture_data[i * 4 + RED] = texture_data[i * 4 + BLUE];
+			texture_data[i * 4 + BLUE] = red;
+		}
+
+		// Almost there, but GDI+'s coordinate system is also upside down 
+		// relative to how we store the pixels.
+		// So, flip the image across the y.
+		for (int y = 0; y < BITMAP_WH / 2; y++) {
+			int y_prime = BITMAP_WH - 1 - y;
+
+			for (int x = 0; x < BITMAP_WH; x++) {
+				int i = x + y * BITMAP_WH;
+				int i_prime = x + y_prime * BITMAP_WH;
+
+				auto red = texture_data[i * 4 + RED];
```

Comment:
Note to self, this could probably be dramatically simplified by using std::swap instead